### PR TITLE
feat(nexus): add support for v1 nexus api calls

### PR DESCRIPTION
**Pull request by Ab-hishek**
- Add v1 nexus api calls(create/destroy/list/publish/unpublish/addchild/removechild)

Signed-off-by: Abhishek Agarwal <abhishek.agarwal@mayadata.io>

### CODE REVIEW

**Comment by Ab-hishek** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors d+

**Comment by Ab-hishek** [comment]
bors try

File: control-plane/agents/core/src/controller/wrapper.rs [hunk 1]
```diff
@@ -1105,17 +1108,36 @@ impl ClientOps for GrpcClientLocked {
...
-                unimplemented!()
+                let rpc_nexus = self
+                    .client_v1()?
+                    .nexus()
+                    .create_nexus(v1_conversion::to_rpc(request))
+                    .await
+                    .context(GrpcRequestError {
+                        resource: ResourceKind::Nexus,
+                        request: "create_nexus",
+                    })?;
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/controller/wrapper.rs:1239]: We need to handle the idempotency differently now - If the nexus exists the create fails on the v1 api.

File: control-plane/agents/core/src/controller/wrapper.rs [hunk 2]
```diff
@@ -1134,7 +1156,16 @@ impl ClientOps for GrpcClientLocked {
...
-                unimplemented!()
+                let _ = self
+                    .client_v1()?
+                    .nexus()
+                    .destroy_nexus(v1_conversion::to_rpc(request))
+                    .await
+                    .context(GrpcRequestError {
+                        resource: ResourceKind::Nexus,
+                        request: "destroy_nexus",
+                    })?;
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/controller/wrapper.rs:1277]: Same here, if it does not exist it fails on the v1 api.
And probably the other ones have changes as well...

File: control-plane/agents/core/src/controller/wrapper.rs (hunk 1 above)
Comment (Ab-hishek) [reply control-plane/agents/core/src/controller/wrapper.rs:1239]: Ohh yes! Will handle those.

File: control-plane/agents/core/src/controller/grpc.rs [hunk 3]
```diff
@@ -100,10 +102,20 @@ impl GrpcContext {
...
+pub(crate) type NexusClient = rpc::v1::nexus::nexus_rpc_client::NexusRpcClient<Channel>;
 
 #[derive(Clone, Debug)]
 pub(crate) struct MayaClientV1 {
     host: HostClient,
+    nexus: NexusClient,
+}
+
+impl MayaClientV1 {
+    /// get the v1 nexus client
```
Comment (tiagolobocastro) [inline control-plane/agents/core/src/controller/grpc.rs]:
```suggestion
    /// Get the v1 nexus client.
```

File: rpc/src/lib.rs [hunk 4]
```diff
@@ -118,4 +118,15 @@ pub mod v1 {
             host_rpc_client, BlockDevice, ListBlockDevicesRequest,
         };
     }
+
+    /// V1 Nexus autogenerated grpc code
```
Comment (tiagolobocastro) [inline rpc/src/lib.rs]:
```suggestion
    /// V1 Nexus autogenerated grpc code.
```

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Ab-hishek** [comment]
bors try

**Comment by Ab-hishek** [comment]
bors try

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by Ab-hishek** [comment]
bors merge

**Pull request merged**